runtime/vm/xz64/UpcallThunkGen.cpp

```diff
@@ -0,0 +1,922 @@
+ * The struct is smaller thani or equal to 4096 bytes.
+ */
+static I_32
+generateSmallStructCopyInstructions(I_8 *instructionPtr, I_32 structSize, I_32 offsetToParameterArea)
+{
+	I_32 totalInstructionSize = 0;
+        // Load R4 with the hiddenParameter address
+        totalInstructionSize += LG(instructionPtr, R4, 0, R15, offsetToParameterArea);
+
+	const I_32 numLargeCopies = structSize / maxMVCCopyLength;
+        const I_32 residue = structSize % maxMVCCopyLength;
+
+	for(int i = 0; i < numLargeCopies; i++) {
+		// MVC D1(L,B1),D2(B2)
+		// The 'L' field is 0-based
+		totalInstructionSize += MVC(instructionPtr + totalInstructionSize, i*maxMVCCopyLength, maxMVCCopyLength - 1, R4, i*maxMVCCopyLength, R2);
+	}
+	//printf("generateStructCopyInstructions_0: numLargeCopies: %d, residue: %d, structSize: %d, offsetParameter: %d\n", numLargeCopies, residue, structSize, offsetToParameterArea);
+	if (residue != 0) {
+		// MVC D1(L,B1),D2(B2)
+		totalInstructionSize += MVC(instructionPtr + totalInstructionSize, numLargeCopies * maxMVCCopyLength, residue - 1, R4, numLargeCopies * maxMVCCopyLength, R2);
+	}
+	return totalInstructionSize;
+}
+
+/*
+ * Generates a sequence of instructions to copy data from a Java MemorySegment Object to a C struct.
+ * The struct is larger than 4096 bytes.
+ */
+static I_32
+generateLargeStructCopyInstructions(I_8 *instructionPtr, I_32 structSize, I_32 offsetToParameterArea)
+{
+	I_32 totalInstructionSize = 0;
+
+	const I_32 numFullSlots = structSize / maxMVCCopyLength;
+        const I_32 residue = structSize % maxMVCCopyLength;
+
+	const I_32 maxLargeCopies = 16;
+	const I_32 loopInstructionsLength = 114; // size in bytes
+	const I_32 sourceRegister = R2;
+	const I_32 destinationRegister = R4;
+	I_32 destinationDisplacementInt = -1;
+	I_32 sourceDisplacementInt = -1;
+	I_32 currMVCLength = -1;
+
+	// Load R4 with the hiddenParameter address
+        totalInstructionSize += LG(instructionPtr, destinationRegister, 0, R15, offsetToParameterArea);
+	// Load loop counter into R3
+	totalInstructionSize += IILF(instructionPtr + totalInstructionSize, R3, numFullSlots);
+
+	for (int i = 0; i < maxLargeCopies; i++) {
+		// MVC D1(L,B1),D2(B2)
+		// The 'L' field is 0-based
+		destinationDisplacementInt = i*maxMVCCopyLength;
+		currMVCLength = maxMVCCopyLength - 1;
+		sourceDisplacementInt = i*maxMVCCopyLength;
+		totalInstructionSize += MVC(instructionPtr + totalInstructionSize, destinationDisplacementInt, currMVCLength, destinationRegister, sourceDisplacementInt, sourceRegister);
+	}
+	//printf("generateStructCopyInstructions_0: numLargeCopies: %d, residue: %d, structSize: %d, offsetParameter: %d\n", numLargeCopies, residue, structSize, offsetToParameterArea);
+	totalInstructionSize += AGFI(instructionPtr + totalInstructionSize, destinationRegister, maxMVCDisplacement);
+	totalInstructionSize += AGFI(instructionPtr + totalInstructionSize, sourceRegister, maxMVCDisplacement);
```

Comment:
From an AGI pipeline bypass perspective, It's better to adjust the internal pointers with `LA/LAY Rdest, maxMVCDisplacement(Rdest)`.